src/helpers/ht_device_info.service.ts

```diff
@@ -70,26 +50,344 @@ class DeviceInfoService {
-        });
+    getDeviceName() {
+        if ( this.browserData && this.browserData.device) {
+            return this.browserData.device.model;
+        } else {
+            return this.getOsType() + ' ' + this.getBrowser() + ' ' + this.getBrowserVersion();
+        }
+
+        return null;
+    }
+
+    getOsType() {
+        if ( this.browserData && this.browserData.os) {
+            return this.browserData.os.name;
+        } else {
+            return this.getUAFromNavigator().os;
+        }
+
+        return null;
+    }
+
+    getBrowser() {
+        if ( this.browserData && this.browserData.browser) {
+            return this.browserData.browser.name;
+        } else {
+            return this.getUAFromNavigator().browserName;
+        }
+
+        return null;
+    }
+
+    getBrowserVersion() {
+        if ( this.browserData && this.browserData.browser) {
+            return this.browserData.browser.version;
+        } else {
+            return this.getUAFromNavigator().fullVersion;
+        }
+
+        return null;
+    }
+
+    getBrowserMajorVersion() {
+        if ( this.browserData && this.browserData.browser) {
+            return this.browserData.browser.version;
+        } else {
+            return this.getUAFromNavigator().majorVersion;
+        }
+
+        return null;
+    }
+
+    getFonts() {
+        return this.fonts;
+    }
+
+    getFingerprint() {
+        if ( !this.fingerprint ) {
+            if ( this.fingerprintComponents && Object.prototype.toString.call( this.fingerprintComponents ) === '[object Array]') {
+                const values = this.fingerprintComponents.map(function (component) { return component.value });
+                this.fingerprint = Fingerprint2.x64hash128(values.join(''), 31);
+                this.pushFPC(this.fingerprintDetails, 'fingerprint', this.fingerprint);
+            } else {
+                return null;
+            }
+        }
+        return this.fingerprint;
+    }
+
+    getFingerprintComponents() {
+
+        if ( !this.fingerprintDetails ) {
+            this.fingerprintDetails = new Array();
+
+            if ( this.fingerprintComponents ) {
+                for ( let i = 0; i < this.fingerprintComponents.length; i++ ) {
+                    if ( this.fingerprintComponents[i] &&
+                        this.fingerprintComponents[i].key !== 'canvas' &&
+                        this.fingerprintComponents[i].key !== 'webgl' ) {
+
+                        if ( this.fingerprintComponents[i].key === 'js_fonts' ) {
+                            this.fonts = this.fingerprintComponents[i].key;
+                        }
+                        this.fingerprintDetails.push({
+                            key: this.fingerprintComponents[i].key,
+                            value: this.joinValues(this.fingerprintComponents[i].value)
+                        });
+                    }
+                }
+
+
+                // FingerprintJS does not generate these
```

Comment:
All extra parameters added from here till the end of the function need to check whether the key already exists in the array, that will impact performance . We could do that only if detailedFingerprint=true

Reply:
We could keep deviceName regardless of the value of detailedFingerprint, that one is important.
30: {key: "deviceName", value: "MacOS Chrome 79.0.3945.117"}

Reply:
detailedFingerprint=true is actually not even working, that was to stop it running ClientJS - introduced as part of messing about with PEXA LP perf, I think
Now that ClientJS is gone, this no longer matters, and the pushFPC checks if field already exists
In short, all fields are populated regardless of detailedFingerprint=true, this needs to be depracated

Reply:
Added to https://www.pivotaltracker.com/story/show/171322089 so we don't delay releasing the security patch